### PR TITLE
feat(job): deploy other version for jobs and advanced settings

### PR DESCRIPTION
**Pull request by bdebon**
# What does this PR do?

- Deploy another version for job with git
- Deploy another version for job with container
- Deploy another version for container
- Advanced Settings

> Link to the JIRA ticket
https://qovery.atlassian.net/jira/software/projects/FRT/boards/23?selectedIssue=FRT-626
https://qovery.atlassian.net/jira/software/projects/FRT/boards/23?selectedIssue=FRT-629

CleanShot 2022-12-29 at 17 36 09@2x

Put screenshot following the template `before` => `after` here

---

### Global

- [x] This PR does not introduce any breaking change
- [ ] This PR introduces breaking change(s) and has been labeled as such
- [x] I have found someone to review this PR and pinged him

### Store

- [ ] This PR introduces new store changes

### NX

- [ ] I have run the dep-graph locally and made sure the tree was clean i.e no circular dependencies
- [ ] I have followed the library pattern i.e `feature`, `ui`, `data`, `utils`

### Clean Code

- [ ] I made sure the code is type safe (no any)
- [ ] I have included a feature flag on my feature, if applicable

### CODE REVIEW

**Comment by evoxmusic** [comment]
A preview environment was automatically created via Qovery.
Click on the link below to follow its deployment and use it.
:point_right: [[PR] staging - feat(job): deploy other version for jobs and advanced settings - 2022-12-29T16:36:01Z](https://console.qovery.com/platform/organization/141c07c8-0dd9-4623-983b-3fdd61867255/projects/977bd069-20f7-4b24-bd95-28b44174cf43/environments/2cee1a6a-438e-4ecc-9851-225cb3d966f0/applications)

 Another comment will be posted when deployments are terminated

**Comment by evoxmusic** [comment]
Your preview environment has been successfully deployed !
Click on the link below to open your service:
:point_right: [console](https://z2cee1a6a-zb25183b2-gtw.za8ad0659.bool.sh)
:point_right: [storybook](https://z2cee1a6a-za9e31aff-gtw.za8ad0659.bool.sh)

**Review by RemiBonnet — COMMENTED** [review]
Thanks @bdebon, here my suggestions to improve it :

File: libs/pages/application/src/lib/feature/page-settings-advanced-feature/page-settings-advanced-feature.spec.tsx [hunk 1]
```diff
@@ -108,42 +111,51 @@ describe('PageSettingsAdvancedFeature', () => {
 
   // I think the useForm hook also use useState, this is why the first 4th call are to ignored. https://gist.github.com/mauricedb/eb2bae5592e3ddc64fa965cde4afe7bc
   it('should set the keys if application and advanced_settings are defined', async () => {
-    mockApplication.advanced_settings!.loadingStatus = 'loaded'
+    ;(mockApplication as GitApplicationEntity).advanced_settings!.loadingStatus = 'loaded'
```
[the remark this answers is not in the document]
Comment (bdebon) [reply libs/pages/application/src/lib/feature/page-settings-advanced-feature/page-settings-advanced-feature.spec.tsx:114]: It is added by prettier and is apparently a security:
https://prettier.io/docs/en/rationale.html#semicolons

**Comment by bdebon** [comment]
- I don't think there is too much `as`. Job as introduced a lot of complexity in the overall typing of the project. We decided to consider it as an Application but it's in the end a weird application than can be either an application or a container. It adds some complexity and some extra typing. I don't think there is such a thing as using too many `as`. I used it when it's needed. I will change the only place I can use your other writing for that but all in all we can't reduce so much the `as` usage unfortunately. And I agree, it makes the code more verbose and harder to read but I don't think we have options here... Feel free to dig if you have time.

- For the application name repeated in the modal, it's a request from Alessandro and something that we're gonna have to implement for almost all modals. 
It's not exactly the final design of this though. But we decided with Alessandro to wait a little bit before implementing the design with Icons.
CleanShot 2022-12-29 at 18 27 56@2x

**Comment by bdebon** [comment]
I finally was able to remove a lot of `as` by implementing a full refactoring of the app by declaring the interface `ApplicationEntity` way differently. Before it was a type that was keeping only the property that were common to each of the property. We changed this idea for this approach:

``` 
export interface ApplicationEntity
  extends Partial<ContainerApplicationEntity>,
    Omit<Partial<GitApplicationEntity>, 'description' | 'default_advanced_settings' | 'advanced_settings'>,
    Partial<JobApplicationEntity> {
  id: string
  created_at: string
  name: string
}
```

It is a little bit more complex to declare but becomes much closer to the abstraction we decided which is to consider Job, Containers and GitApplication as an ApplicationEntity object. It drastically reduced the need for casting everywhere.
I had to move and adapt many files but it's a good win for the future! I'm super happy with this solution, it something I had in mind for a long time now but I did not know how to write it well...

**Pull request merged**